File: packages/backstage-ts/test/cron.test.ts
/**
 * Cron parsing edge cases
 */
import { describe, test, expect } from 'bun:test';
import { CronTask } from '../src/cron';

describe('CronTask Edge Cases', () => {
  test('every minute', () => {
    const task = new CronTask('* * * * *', 'every.minute');
    const now = new Date('2024-01-10T12:00:00Z');
    const next = task.getNextRun(now);
    expect(next.getMinutes()).toBe(1);
  });

  test('every 5 minutes', () => {
    const task = new CronTask('*/5 * * * *', 'every.5.minutes');
    const now = new Date('2024-01-10T12:03:00Z');
    const next = task.getNextRun(now);
    expect(next.getMinutes()).toBe(5);
  });

  test('every hour', () => {
    const task = new CronTask('0 * * * *', 'every.hour');
    const now = new Date('2024-01-10T12:30:00Z');
    const next = task.getNextRun(now);
    expect(next.getMinutes()).toBe(0);
    expect(next.getHours()).toBe(13);
  });

  test('daily at midnight', () => {
    const task = new CronTask('0 0 * * *', 'daily.midnight');
    const now = new Date('2024-01-10T12:00:00Z');
    const next = task.getNextRun(now);
    expect(next.getHours()).toBe(0);
    expect(next.getMinutes()).toBe(0);
  });

  test('weekdays only', () => {
    const task = new CronTask('0 9 * * 1-5', 'weekdays.9am');
    const friday = new Date('2024-01-12T08:00:00Z'); // Friday
    const next = task.getNextRun(friday);
    expect(next.getDay()).toBeGreaterThanOrEqual(1);
    expect(next.getDay()).toBeLessThanOrEqual(5);
  });

  test('weekends only', () => {
    const task = new CronTask('0 10 * * 0,6', 'weekends');
    const wednesday = new Date('2024-01-10T08:00:00Z');
    const next = task.getNextRun(wednesday);
    expect([0, 6]).toContain(next.getDay());
  });

  test('specific day of month', () => {
    const task = new CronTask('0 0 15 * *', 'monthly.15th');
    const now = new Date('2024-01-10T12:00:00Z');
    const next = task.getNextRun(now);
    expect(next.getDate()).toBe(15);
  });

  test('specific months', () => {
    const task = new CronTask('0 0 1 1,7 *', 'jan.jul');
    const now = new Date('2024-03-01T00:00:00Z');
    const next = task.getNextRun(now);
    expect([1, 7]).toContain(next.getMonth() + 1);
  });

  test('complex schedule', () => {
    const task = new CronTask('30 9-17 * * 1-5', 'business.hours');
    const now = new Date('2024-01-10T08:00:00Z'); // Wednesday 8am
    const next = task.getNextRun(now);
    expect(next.getMinutes()).toBe(30);
    expect(next.getHours()).toBeGreaterThanOrEqual(9);
    expect(next.getHours()).toBeLessThanOrEqual(17);
  });

  test('step in range', () => {
    const task = new CronTask('0-30/10 * * * *', 'step.range');
    const now = new Date('2024-01-10T12:05:00Z');
    const next = task.getNextRun(now);
    expect([0, 10, 20, 30]).toContain(next.getMinutes());
  });

  test('throws on invalid schedule - wrong field count', () => {
    expect(() => new CronTask('* * *', 'invalid')).toThrow();
  });

  test('throws on invalid schedule - bad step', () => {
    expect(() => new CronTask('*/0 * * * *', 'zero.step')).toThrow();
  });

  test('throws on invalid schedule - invalid value', () => {
    expect(() => new CronTask('abc * * * *', 'letters')).toThrow();
  });

  test('marks run correctly', () => {
    const task = new CronTask('* * * * *', 'test');
    const now = new Date();
    task.markRun(now);
    // Should calculate next from marked time
    const next = task.getNextRun(now);
    expect(next > now).toBe(true);
  });

  test('handles year boundary', () => {
    const task = new CronTask('0 0 1 1 *', 'new.year');
    const now = new Date('2024-12-31T23:00:00Z');
    const next = task.getNextRun(now);
    expect(next.getFullYear()).toBe(2025);
  });
});

File: packages/backstage-go/cron_test.go
package backstage

import (
	"testing"
	"time"
)

func TestCronEveryMinute(t *testing.T) {
	task, err := NewCronTask("* * * * *", "every.minute", nil)
	if err != nil {
		t.Fatal(err)
	}

	now := time.Date(2024, 1, 10, 12, 0, 0, 0, time.UTC)
	next := task.NextRun(now)

	if next.Minute() != 1 {
		t.Errorf("expected minute 1, got %d", next.Minute())
	}
}

func TestCronEveryFiveMinutes(t *testing.T) {
	task, _ := NewCronTask("*/5 * * * *", "every.5.min", nil)
	now := time.Date(2024, 1, 10, 12, 3, 0, 0, time.UTC)
	next := task.NextRun(now)

	if next.Minute()%5 != 0 {
		t.Errorf("expected multiple of 5, got %d", next.Minute())
	}
}

func TestCronEveryHour(t *testing.T) {
	task, _ := NewCronTask("0 * * * *", "every.hour", nil)
	now := time.Date(2024, 1, 10, 12, 30, 0, 0, time.UTC)
	next := task.NextRun(now)

	if next.Minute() != 0 {
		t.Errorf("expected minute 0, got %d", next.Minute())
	}
	if next.Hour() != 13 {
		t.Errorf("expected hour 13, got %d", next.Hour())
	}
}

func TestCronDailyMidnight(t *testing.T) {
	task, _ := NewCronTask("0 0 * * *", "daily.midnight", nil)
	now := time.Date(2024, 1, 10, 12, 0, 0, 0, time.UTC)
	next := task.NextRun(now)

	if next.Hour() != 0 || next.Minute() != 0 {
		t.Errorf("expected 00:00, got %02d:%02d", next.Hour(), next.Minute())
	}
}

func TestCronBusinessHours(t *testing.T) {
	task, _ := NewCronTask("0 9-17 * * 1-5", "business", nil)
	now := time.Date(2024, 1, 10, 8, 0, 0, 0, time.UTC) // Wednesday 8am
	next := task.NextRun(now)

	if next.Hour() < 9 || next.Hour() > 17 {
		t.Errorf("expected 9-17, got %d", next.Hour())
	}
	if next.Weekday() < 1 || next.Weekday() > 5 {
		t.Errorf("expected weekday, got %v", next.Weekday())
	}
}

func TestCronWeekends(t *testing.T) {
	task, _ := NewCronTask("0 10 * * 0,6", "weekends", nil)
	now := time.Date(2024, 1, 10, 8, 0, 0, 0, time.UTC) // Wednesday
	next := task.NextRun(now)

	day := int(next.Weekday())
	if day != 0 && day != 6 {
		t.Errorf("expected weekend, got %v", next.Weekday())
	}
}

func TestCronSpecificDayOfMonth(t *testing.T) {
	task, _ := NewCronTask("0 0 15 * *", "monthly.15th", nil)
	now := time.Date(2024, 1, 10, 0, 0, 0, 0, time.UTC)
	next := task.NextRun(now)

	if next.Day() != 15 {
		t.Errorf("expected day 15, got %d", next.Day())
	}
}

func TestCronStepInRange(t *testing.T) {
	task, _ := NewCronTask("0-30/10 * * * *", "step.range", nil)
	now := time.Date(2024, 1, 10, 12, 5, 0, 0, time.UTC)
	next := task.NextRun(now)

	valid := []int{0, 10, 20, 30}
	found := false
	for _, v := range valid {
		if next.Minute() == v {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("expected one of %v, got %d", valid, next.Minute())
	}
}

func TestCronYearBoundary(t *testing.T) {
	task, _ := NewCronTask("0 0 1 1 *", "new.year", nil)
	now := time.Date(2024, 12, 31, 23, 0, 0, 0, time.UTC)
	next := task.NextRun(now)

	if next.Year() != 2025 {
		t.Errorf("expected 2025, got %d", next.Year())
	}
	if next.Month() != 1 || next.Day() != 1 {
		t.Errorf("expected Jan 1, got %v", next)
	}
}

func TestCronInvalidSchedules(t *testing.T) {
	testCases := []struct {
		schedule string
		desc     string
	}{
		{"invalid", "single word"},
		{"* * *", "3 fields"},
		{"* * * * * *", "6 fields"},
		{"60 * * * *", "minute 60"},
		{"* 24 * * *", "hour 24"},
		{"* * 32 * *", "day 32"},
		{"* * * 13 *", "month 13"},
		{"* * * * 7", "weekday 7"},
		{"*/0 * * * *", "step 0"},
		{"abc * * * *", "letters"},
	}

	for _, tc := range testCases {
		_, err := NewCronTask(tc.schedule, "test", nil)
		if err == nil {
			t.Errorf("expected error for %s (%s)", tc.schedule, tc.desc)
		}
	}
}

func TestCronMarkRun(t *testing.T) {
	task, _ := NewCronTask("* * * * *", "test", nil)
	now := time.Now()
	task.MarkRun(now)

	if task.lastRun != now {
		t.Error("lastRun not set")
	}
}
